src/api/app/jobs/cleanup_events.rb

```diff
@@ -1,7 +1,23 @@
 class CleanupEvents < ApplicationJob
   def perform
     Event::Base.transaction do
-      Event::Base.where(project_logged: true, queued: true, undone_jobs: 0).lock(true).delete_all
+      events_to_be_deleted.lock(true).delete_all
     end
   end
+
+  private
+
+  def events_to_be_deleted
+    # delete_all will not work with this query so we need to find again by id
+    events_to_be_deleted_array =
+      Event::Base.left_outer_joins(:digest_emails)
+        .select('events.*, MIN(digest_emails.email_sent) AS has_all_digest_emails_sent')
+        .where(project_logged: true, queued: true, undone_jobs: 0)
+        .group('events.id')
+        .having('has_all_digest_emails_sent = 1 OR has_all_digest_emails_sent IS NULL')
+
+    event_ids = events_to_be_deleted_array.map(&:id)
```

Comment:
@coolo I think you've misunderstood this a bit, or at least  your question doesn't really make sense to me? Because the more power users that turn on digest emails, the less events there will be in `events_to_be_deleted_array` here.

What this query does it to find all events that would be deleted but exclude events that have one or more digest_emails which have not been sent. Because we still need the events in the database in order to send the digest email at the end of the day.

I'm going to add a comment explaining this because the query is complex..